### PR TITLE
eth/filters: subscribe history logs

### PR DESCRIPTION
**Pull request by jsvisa**
This is the second part of #15063

### CODE REVIEW

File: eth/filters/api.go [hunk 1]
```diff
@@ -258,29 +258,131 @@ func (api *FilterAPI) Logs(ctx context.Context, crit FilterCriteria) (*rpc.Subsc
 		matchedLogs = make(chan []*types.Log)
 	)
 
-	logsSub, err := api.events.SubscribeLogs(ethereum.FilterQuery(crit), matchedLogs)
-	if err != nil {
-		return nil, err
+	filterLiveLogs := func() error {
```
[the remark this answers is not in the document]
Comment (jsvisa) [reply eth/filters/api.go:266]: thanks for the advice, declare them as private methods.

File: eth/filters/api.go [hunk 2]
```diff
@@ -258,29 +258,131 @@ func (api *FilterAPI) Logs(ctx context.Context, crit FilterCriteria) (*rpc.Subsc
...
 
+	n := crit.FromBlock.Int64()
+	go func() {
+		// do historical sync first
+		head, err := filterHistLogs(n)
+		if err != nil {
+			return
+		}
+		// if toBlock is limited and handled, no need to subscribe live logs anymore
+		if toBlock := crit.ToBlock; toBlock != nil && toBlock.Sign() > 0 && toBlock.Cmp(big.NewInt(head)) <= 0 {
```
[the remark this answers is not in the document]
Comment (jsvisa) [reply eth/filters/api.go:386]: eg: someone is only interested in `[header - 100k, header + 1000]`(he also knows the current head what is), after processing the hist logs, the current header is moving to `old header + 1001`, so the live is not needed.

[the remark this answers is not in the document]
Comment (s1na) [reply eth/filters/api.go:389]: As we discussed offline let's leave subscribing to a limited log range (i.e. historical logs but no live) feature for a future PR.

**Comment by s1na** [comment]
I have some misgivings about how we handle block ranges right now. IMO block ranges specifically for the live mode are not necessary. I.e. when you subscribe you get the logs for all new blocks. To make my case [web3.js](https://web3js.readthedocs.io/en/v1.2.11/web3-eth-subscribe.html) only lets you specify FromBlock (no ToBlock) and [ethers.js](https://docs.ethers.org/v5/api/providers/provider/#Provider--events) doesn't support any of those parameters. These are 2 of the biggest web3 libraries.

My suggestion is we can simplify this greatly by: only allowing `FromBlock` and only for specifying historical blocks. I.e. `FromBlock` should be either a number below head block or `safe` or `finalized`. Later on we can allow `ToBlock` (for historical range queries).

**Comment by jsvisa** [comment]
@s1na lgtm, and as we discussed offline, seems we have more work to do, I'll post them here(in case I forget).

- [ ] support block range subscription;
- [ ] support server active close connection;

**Comment by fjl** [comment]
I like @s1na's idea to not allow `toBlock` for subscription queries. The problem with using our subscription model is that there is no way for the server to signal the end of a subscription. We can fix that by changing how subscriptions work, but I think that will be a lot of work for everyone to support. Let's try the simple solution for now.

File: eth/filters/api.go [hunk 3]
```diff
@@ -269,19 +330,94 @@ func (api *FilterAPI) Logs(ctx context.Context, crit FilterCriteria) (*rpc.Subsc
...
+		if n >= head {
+			return head, nil
+		}
+
+		// do historical sync from n to head
+		f := api.sys.NewRangeFilter(n, head, crit.Addresses, crit.Topics)
+		logChan, errChan := f.rangeLogsAsync(cctx)
+
+		// subscribe rmLogs
+		query := ethereum.FilterQuery{FromBlock: big.NewInt(n), ToBlock: big.NewInt(head), Addresses: crit.Addresses, Topics: crit.Topics}
```
Comment (s1na) [inline eth/filters/api.go:359]: Took me a minute to get this, but smart trick I love it!

Comment (s1na) [reply eth/filters/api.go:359]: Reality check: It's more complicated. 

- Reorg happens in a future block that `rangeLogsAsync` hasn't processed yet. In that case we don't really care about the event since when `rangeLogsAsync` reaches that point it will go on the right chain by itself.
- We can keep track of how far we are in delivery by checking the block number of incoming logs.
- However it's harder to say where is the point of reorg, because removed logs are sent in the reverse order (from more recent block backwards).
- What we can do is compare as they come in to see if they pass the blocks for which we've already delivered logs.
- At this point we have to stop `rangeAsyncLogs` immediately. Or hmm there's a question what should be done if it is mid-block (with some logs remaining in that block).
- Then the new subscription which sends removed logs will also send the replacement logs of the new chain. So we can just tune into that.

File: eth/filters/api.go [hunk 4]
```diff
@@ -355,44 +357,71 @@ func (api *FilterAPI) histLogs(notifier notifier, rpcSub *rpc.Subscription, from
...
-				rmLogsSub.Unsubscribe()
+				reorgLogsSub.Unsubscribe()
 				return 0, err
 			case log := <-logChan:
-				notifier.Notify(rpcSub.ID, &log)
-			case logs := <-rmLogsCh:
+				// We transmit all data that meets the following conditions:
+				// 1. reorg not happened
+				// 2. reorg happened but the the log is in the remainder of the currently delivered block
+				if !reorged || log.BlockNumber < reorgBlock || log.BlockNumber <= delivered {
```
Comment (s1na) [inline eth/filters/api.go:393]: I don't understand this condition: `log.BlockNumber <= delivered`. Why did you add this?

File: eth/filters/api.go [hunk 5]
```diff
@@ -355,44 +357,71 @@ func (api *FilterAPI) histLogs(notifier notifier, rpcSub *rpc.Subscription, from
...
+				} else {
+					// No more rangelogs are needed, let's stop the rangeLogsAsync
+					cancel()
+				}
+			case logs := <-reorgLogsCh:
 				for _, log := range logs {
-					log := log
-					if log.Removed {
-						rmLogs = append(rmLogs, log)
+					reorgLogs = append(reorgLogs, log)
```
Comment (s1na) [inline eth/filters/api.go:402]: This doesn't seem right. We're keeping the logs (and later emitting them) even if it turns out we don't care about it (e.g. reorg happened in future).

File: eth/filters/api.go [hunk 6]
```diff
@@ -355,44 +357,71 @@ func (api *FilterAPI) histLogs(notifier notifier, rpcSub *rpc.Subscription, from
...
+					// Update the reorgBlock to the last removed log's block number or the first added log's
+					if reorgBlock == 0 || log.Removed {
```
Comment (s1na) [inline eth/filters/api.go:404]: Hm here it seems you're anticipating the case that there are no removed logs, only "replaced" logs 🤔 how could this happen?

File: eth/filters/api.go (hunk 4 above)
Comment (jsvisa) [reply eth/filters/api.go:393]: Ensure that all logs of the same block are sent out, avoid sending only a subset of logs.

File: eth/filters/api.go (hunk 5 above)
Comment (jsvisa) [reply eth/filters/api.go:402]: yes, if the reorg happens in the future, the reorgLogs will be reset. https://github.com/ethereum/go-ethereum/pull/27439/files#diff-bbe984989be76dbc39ff9214d3a13800c4f876328ca26e98242db6e6774b3761R411

File: eth/filters/api.go (hunk 6 above)
Comment (jsvisa) [reply eth/filters/api.go:404]: Maybe a block doesn't contain any logs, or even an empty block without any transactions included

Comment (s1na) [reply eth/filters/api.go:404]: Good call

File: eth/filters/api.go (hunk 5 above)
Comment (s1na) [reply eth/filters/api.go:402]: Doesn't that simply make a shallow copy of the slice? I don't see how it resets it.

File: eth/filters/api.go [hunk 7]
```diff
@@ -269,19 +319,123 @@ func (api *FilterAPI) Logs(ctx context.Context, crit FilterCriteria) (*rpc.Subsc
...
+					cancel()
+					continue
+				}
+				delivered = log.BlockNumber
+				notifier.Notify(rpcSub.ID, &log)
+			case logs := <-reorgLogsCh:
+				if len(logs) == 0 {
+					continue
+				}
+				if reorgBlock == 0 {
```
[the remark this answers is not in the document]
Comment (s1na) [reply eth/filters/api.go:402]: Yes you're right that considering 2 different reorgs adds another layer of complexity. I have not considered that in the changes I made.

Regarding the scenario. Yes I think we should disable the `errChan` breaking out of the loop if we're in the middle of a reorg processing. But I also don't know how to detect the end of a reorg yet.

File: eth/filters/api.go [hunk 8]
```diff
@@ -269,19 +310,164 @@ func (api *FilterAPI) Logs(ctx context.Context, crit FilterCriteria) (*rpc.Subsc
...
+
+		// Do historical sync beginning at `from` to head.
+		f := api.sys.NewRangeFilter(from, head, crit.Addresses, crit.Topics)
+		logsCh, errChan := f.rangeLogsAsync(ctx)
+
+	FORLOOP:
+		for {
+			select {
+			case <-closeC:
+				return nil
```
Comment (s1na) [inline eth/filters/api.go:455]: We should cancel the range filter's context here.

File: eth/filters/api.go [hunk 9]
```diff
@@ -280,8 +321,175 @@ func (api *FilterAPI) Logs(ctx context.Context, crit FilterCriteria) (*rpc.Subsc
 			}
 		}
 	}()
+	return nil
+}
 
-	return rpcSub, nil
+// histLogs retrieves logs older than current header.
+func (api *FilterAPI) histLogs(notifier notifier, rpcSub *rpc.Subscription, from int64, crit FilterCriteria) error {
```
Comment (holiman) [inline eth/filters/api.go:328]: I think `histLogs` is a bit wonky, is the benefit of reducing `historicalLogs` really worth it?

File: eth/filters/api.go [hunk 10]
```diff
@@ -246,31 +251,67 @@ func (api *FilterAPI) NewHeads(ctx context.Context) (*rpc.Subscription, error) {
...
+// The following criteria are valid:
+// * from: nil, to: nil -> yield live logs.
+// * from: blockNum | safe | finalized, to: nil -> historical beginning at `from` to head, then live logs.
+// * Every other case should fail with an error.
+func (api *FilterAPI) logs(ctx context.Context, notifier notifier, rpcSub *rpc.Subscription, crit FilterCriteria) error {
+	if crit.ToBlock != nil {
+		return errInvalidToBlock
+	}
+	if crit.FromBlock == nil {
+		return api.liveLogs(notifier, rpcSub, crit)
```
Comment (holiman) [inline eth/filters/api.go:283]: on input, you take a `ctx`. But the ctx is lost, afaict you don't check for ctx timeouts or cancel. Shouldn't you? What is the lifecycle here?

Comment (jsvisa) [reply eth/filters/api.go:283]: As the `ctx` will be canceled when the subscription goroutine returns, so we instantiated a new background context to control the workflow. 

The lifecycle is as below:

1. do historical logs fetch and push;
2. do live logs push;
3. terminate if the subscription was canceled or the push channel was broken;

File: eth/filters/api.go [hunk 11]
```diff
@@ -280,8 +321,178 @@ func (api *FilterAPI) Logs(ctx context.Context, crit FilterCriteria) (*rpc.Subsc
...
+					continue
+				}
+				if liveOnly {
+					notifyLogsIf(notifier, rpcSub.ID, logs, nil)
+					continue
+				}
+				reorgBlock := logs[0].BlockNumber
+				if !reorged && reorgBlock <= delivered {
+					logger.Info("Reorg detected", "reorgBlock", reorgBlock, "delivered", delivered)
+					reorged = true
```
Comment (holiman) [inline eth/filters/api.go:389]: This seems a bit strange. Once `reorged` i sset, it never becomes unset again. It will discard all `histLogs` when it's in that mode. Seems to me that this big switch has thee different modes: 

- reorged mode, where it discards historic logs
- liveOnly mode
- "normal" non-reorged mode

It makes it a bit complicated to follow, IMO

Comment (jsvisa) [reply eth/filters/api.go:389]: > Once reorged i sset, it never becomes unset again

Here we only need to handle the case when our historical delivery process is catching up with the live blocks.

> liveOnly mode

Yeah, We did have another switch `liveOnly`, which was set after all historical logs were delivered: https://github.com/ethereum/go-ethereum/blob/633c7ac67a6de35e000bdc72d51f1054f4e743b5/eth/filters/api.go#L373-L375

Comment (s1na) [reply eth/filters/api.go]: > Seems to me that this big switch has thee different modes:

Actually I managed to bring it down to 2 modes (https://github.com/ethereum/go-ethereum/pull/27439/commits/674a888b060b6acabff0cdcd3ab12fc8fbecb5fd). Now only setting `liveOnly` mode. This will be set:

- either when a reorg happens to discard historical logs
- or when historical logs have finished

But my patch has a problem. In the "reorg" mode, we only send removed log notifications up to the `delivered` point. So if we finish historical processing (say head = 10 so delivered = 10). Then after emitting block 11, we get a reorg from block 8. We will only delivered the removed logs up to 10.

Comment (s1na) [reply eth/filters/api.go]: Ok I pushed another commit which should fix the problem above and IMO the code is more readable now.

That said I noticed the tests fail every once in a while. I believe the edge case they're catching is the following:

- Blockchain will first write the new (reorged) blocks to db and *then* send removed/new logs. This can cause the history processor to go accidentally on the new chain before we realize there's a reorg happening.
- So we return `oldBlock1, oldBlock2, newBlock3, removeOldBlock3, newBlock3, newBlock4`. I.e. the removed logs don't match the ones we emitted for that height.

Comment (s1na) [reply eth/filters/api.go]: I've pushed a fix for this. The subscription now tracks a number of block hashes that it has delivered logs for. This map is used to avoid sending removed logs for a block we haven't delivered as well as sending duplicate logs.

File: eth/filters/api.go [hunk 12]
```diff
@@ -280,8 +328,225 @@ func (api *FilterAPI) Logs(ctx context.Context, crit FilterCriteria) (*rpc.Subsc
...
+			if batch.removed && !hashes.Contains(batch.hash) {
+				continue
+			}
+			if !batch.removed && hashes.Contains(batch.hash) {
+				continue
+			}
+			hashes.Add(batch.hash, struct{}{})
+		}
+		for _, log := range logs[batch.start:batch.end] {
+			log := log
```
Comment (MariusVanDerWijden) [inline eth/filters/api.go:546]: Do we need to deref the log here to copy it? 
e.g. `log := *log`
Otherwise we just copy the pointer if I understand correctly

File: eth/filters/filter_system_test.go [hunk 13]
```diff
@@ -957,3 +973,524 @@ func flattenLogs(pl [][]*types.Log) []*types.Log {
...
+type mockNotifier struct {
+	c chan interface{}
+}
+
+func newMockNotifier() *mockNotifier {
+	return &mockNotifier{c: make(chan interface{})}
+}
+
+func (n *mockNotifier) Notify(id rpc.ID, data interface{}) error {
+	n.c <- data
```
Comment (fjl) [inline eth/filters/filter_system_test.go:986]:
```
log := data.(*types.Log)
n.callback(log)
n.logs = append(n.logs, log)
```

File: eth/filters/filter_system_test.go [hunk 14]
```diff
@@ -957,3 +973,524 @@ func flattenLogs(pl [][]*types.Log) []*types.Log {
...
+	return nil
+}
+
+func (n *mockNotifier) Closed() <-chan interface{} {
+	return nil
+}
+
```
Comment (fjl) [inline eth/filters/filter_system_test.go:993]:
```
func (n *mockNotifier) Done() {
   close(n.done)
}
```

and call `Done` from FilterAPI.

File: eth/filters/api.go [hunk 15]
```diff
@@ -280,8 +328,225 @@ func (api *FilterAPI) Logs(ctx context.Context, crit FilterCriteria) (*rpc.Subsc
...
+	go func() {
+		histDone <- api.doHistLogs(ctx, from, crit.Addresses, crit.Topics, histLogs)
+	}()
+
+	// Compose and notify the logs from liveLogs and histLogs
+	go func() {
+		defer func() {
+			liveLogsSub.Unsubscribe()
+			cancel()
+		}()
```
Comment (fjl) [inline eth/filters/api.go:366]: `defer notifier.Done()`

**Comment by cong08** [comment]
why not merge it ?

**Comment by cong08** [comment]
> why not merge it ?

@karalabe @rjl493456442

**Comment by rjl493456442** [comment]
@jsvisa Please rebase it, we still want it!

**Comment by jsvisa** [comment]
> @jsvisa Please rebase it, we still want it!

Sure, I'll rebase it from master later

**Comment by jsvisa** [comment]
Rebased, but still need some time to test it :)

**Comment by FournyP** [comment]
Hello everyone, any news about this? Faced a issue related to that today
